### PR TITLE
Add device crud

### PR DESCRIPTION
**Pull request by vikaluk**
**STATUS**
READY/IN DEVELOPMENT

**DESCRIPTION**

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [ ] code coverage >90%
- [x] DB indexes added if needed
- [ ] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [ ] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs
- [x] code checked with Rubocop and offenses are fixed

image
image
image

### CODE REVIEW

File: app/controllers/account/devices_controller.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
+class Account::DevicesController < Account::DashboardController
+  def index
+    @devices = collection    
+  end
+
+  def new
+    @device = Device.new
```
Comment (romaslmd) [inline app/controllers/account/devices_controller.rb]: `parent.devices.build`

File: app/controllers/account/devices_controller.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  def collection
+    parent.devices
+  end
+
+  def device_params
+    params.require(:device).permit(:name, :device_type)
+  end
+
+  def resource
+    parent.devices.find(params[:id])
```
Comment (igorkasyanchuk) [inline app/controllers/account/devices_controller.rb]: collection.find(params[:id])

File: db/migrate/20190529183918_create_devices.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class CreateDevices < ActiveRecord::Migration[5.2]
+  def change
+    create_table :devices do |t|
+      t.string :name
+      t.string :device_type
+      t.integer :device_group_id
```
Comment (romaslmd) [inline db/migrate/20190529183918_create_devices.rb]: ?????

File: db/migrate/20190530080128_add_device_organization_id.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddDeviceOrganizationId < ActiveRecord::Migration[5.2]
+  def change
+    add_reference :devices, :organization, foreign_key: true
```
Comment (romaslmd) [inline db/migrate/20190530080128_add_device_organization_id.rb]: why do u need a separate migration?

File: db/migrate/20190530082118_remove_column_devices.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class RemoveColumnDevices < ActiveRecord::Migration[5.2]
+  def change
+    remove_column :devices, :device_group_id
```
Comment (romaslmd) [inline db/migrate/20190530082118_remove_column_devices.rb]: ???????

File: spec/controllers/account/devices_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  end
+
+  describe 'GET #edit' do
+    let!(:user) { create(:user, own_organization: organization) }
+    let!(:organization) { create(:organization_with_device) }
+
+    context 'with signed in user' do
+      before(:each) do
+        sign_in user
+        get :edit, params: { id: user.own_organization.devices.first.id }
```
Comment (romaslmd) [inline spec/controllers/account/devices_controller_spec.rb]: where does this come from?
test should be explicit, to understand where this or that data come from and how it should behave

File: spec/controllers/account/devices_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    end
+  end
+
+  describe '#create' do
+    let!(:user) { create(:user) }
+    let!(:device) { create(:device) }
+
+    context 'as an authenticated user' do
+      before do
+        @device_params = FactoryBot.attributes_for(:device)
```
Comment (romaslmd) [inline spec/controllers/account/devices_controller_spec.rb]: why instance var?

File: spec/controllers/account/devices_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+      it 'create device' do
+        sign_in user
+        expect do
+          post :create, params: { device: @device_params }
+        end.to change(user.own_organization.devices, :count).by(1)
+      end
+
+      it 'it should redirect after save' do
+        sign_in user
```
Comment (romaslmd) [inline spec/controllers/account/devices_controller_spec.rb]:
```
let(:valid_params) { attributes_for(...) }

before(:each) do
  sign_in(user) # <<<< required for each example..
end
```

File: spec/controllers/account/devices_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  describe '#update' do
+    let!(:user) { create(:user, own_organization: organization) }
+    let!(:organization) { create(:organization_with_device) }
+
+    context 'as an authenticated user' do
+      it 'update device' do
+        new_device_params = FactoryBot.attributes_for(:device)
+        new_device_params[:name] = 'Updated'
+        sign_in user
+        put :update, params: { id: user.own_organization.devices.first.id, device: new_device_params }
```
Comment (romaslmd) [inline spec/controllers/account/devices_controller_spec.rb]: again this chaining..

File: spec/controllers/account/devices_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        user.own_organization.devices.first.reload
+        expect(user.own_organization.devices.first.name).to eq 'Updated'
```
Comment (romaslmd) [inline spec/controllers/account/devices_controller_spec.rb]: first... first.. first...

File: spec/controllers/account/devices_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    let!(:user) { create(:user, own_organization: organization) }
+    let!(:organization) { create(:organization_with_device) }
+
+    context 'with signed in user' do
+      before(:each) do
+        sign_in user
+      end
+
+      it 'create device' do
+        sign_in user
```
Comment (romaslmd) [inline spec/controllers/account/devices_controller_spec.rb]: duplicate!
u already have this in before example!
pay attention!

File: spec/models/device_spec.rb [hunk 12]
```diff
@@ -0,0 +1,19 @@
+require 'rails_helper'
+
+RSpec.describe Device, type: :model do
+  context 'validations tests' do
+    let!(:device_group) { create(:device_group) }
+
+    describe '#name' do
```
Comment (romaslmd) [inline spec/models/device_spec.rb]: interesting approach 👍

File: spec/models/device_spec.rb [hunk 13]
```diff
@@ -0,0 +1,19 @@
+require 'rails_helper'
+
+RSpec.describe Device, type: :model do
+  context 'validations tests' do
+    let!(:device_group) { create(:device_group) }
```
Comment (romaslmd) [inline spec/models/device_spec.rb]: no need to create this! it's never used below

File: spec/factories/organizations.rb [hunk 14]
```diff
@@ -7,5 +7,11 @@
         create(:device_group, organization: organization)
       end
     end
+
+    factory :organization_with_device do
```
Comment (romaslmd) [inline spec/factories/organizations.rb]: what's a name is this?
`create(:organization, :organization_with_device)` ? looks and sounds weird

Comment (romaslmd) [reply spec/factories/organizations.rb]: organization may have many devices..

**Pull request merged**